File: crm/src/main/java/serp/project/crm/ui/controller/CustomerController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.crm.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import serp.project.crm.core.domain.dto.PageRequest;
import serp.project.crm.core.domain.dto.request.CreateCustomerRequest;
import serp.project.crm.core.domain.dto.request.CustomerFilterRequest;
import serp.project.crm.core.domain.dto.request.UpdateCustomerRequest;
import serp.project.crm.core.usecase.ActivityUseCase;
import serp.project.crm.core.usecase.CustomerUseCase;
import serp.project.crm.kernel.utils.AuthUtils;

@RestController
@RequestMapping("/api/v1/customers")
@RequiredArgsConstructor
@Slf4j
public class CustomerController {

    private final CustomerUseCase customerUseCase;
    private final ActivityUseCase activityUseCase;

    private final AuthUtils authUtils;

    @PostMapping
    public ResponseEntity<?> createCustomer(@Valid @RequestBody CreateCustomerRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        var response = customerUseCase.createCustomer(request, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/{id}")
    public ResponseEntity<?> updateCustomer(
            @PathVariable Long id,
            @Valid @RequestBody UpdateCustomerRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        var response = customerUseCase.updateCustomer(id, request, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getCustomerById(@PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        var response = customerUseCase.getCustomerById(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{id}/activities")
    public ResponseEntity<?> getActivitiesByCustomerId(
            @PathVariable Long id,
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "20") Integer size) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        PageRequest pageRequest = PageRequest.builder()
                .page(page)
                .size(size)
                .build();

        var response = activityUseCase.getActivitiesByCustomer(id, tenantId, pageRequest);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getAllCustomers(
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "20") Integer size) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }
        PageRequest pageRequest = PageRequest.builder()
                .page(page)
                .size(size)
                .build();

        var response = customerUseCase.getAllCustomers(tenantId, pageRequest);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/search")
    public ResponseEntity<?> filterCustomers(@RequestBody(required = false) CustomerFilterRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        CustomerFilterRequest safeRequest = request != null ? request : CustomerFilterRequest.builder().build();

        var response = customerUseCase.filterCustomers(safeRequest, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteCustomer(@PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        var response = customerUseCase.deleteCustomer(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }
}

File: api_gateway/src/ui/controller/crm/customer_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	"github.com/serp/api-gateway/src/core/service/crm"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type CustomerController struct {
	customerService crm.ICustomerService
}

func (c *CustomerController) CreateCustomer(ctx *gin.Context) {
	var payload map[string]any
	if err := ctx.ShouldBindJSON(&payload); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := c.customerService.CreateCustomer(ctx.Request.Context(), payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *CustomerController) UpdateCustomer(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	var payload map[string]any
	if err := ctx.ShouldBindJSON(&payload); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := c.customerService.UpdateCustomer(ctx.Request.Context(), customerId, payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *CustomerController) GetCustomerByID(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	res, err := c.customerService.GetCustomerByID(ctx.Request.Context(), customerId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *CustomerController) GetCustomers(ctx *gin.Context) {
	page, size, valid := utils.ValidatePaginationParams(ctx)
	if !valid {
		return
	}

	res, err := c.customerService.GetCustomers(ctx.Request.Context(), page, size)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *CustomerController) FilterCustomers(ctx *gin.Context) {
	payload := map[string]any{}
	if ctx.Request.ContentLength != 0 {
		if err := ctx.ShouldBindJSON(&payload); err != nil {
			utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
			return
		}
	}

	res, err := c.customerService.FilterCustomers(ctx.Request.Context(), payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *CustomerController) DeleteCustomer(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	res, err := c.customerService.DeleteCustomer(ctx.Request.Context(), customerId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func NewCustomerController(customerService crm.ICustomerService) *CustomerController {
	return &CustomerController{customerService: customerService}
}
